theseus/src/api/mod_extraction.rs

```diff
@@ -0,0 +1,76 @@
+use std::fs::{self, File};
+use std::path::Path;
+use std::io::BufReader;
+use std::time::SystemTime;
+use serde_json::Value;
+use zip::ZipArchive;
+
+use crate::model::mod_type::JARLoadedMod;
+use crate::util::fetch::sha1_async;
+
+/// Extract meta-data from JAR file.
+#[tracing::instrument]
+pub async fn extract_info_from_jar(file_path: String) -> crate::Result<JARLoadedMod> {
+    let path = Path::new(&file_path);
+
+    let jar_buf = std::fs::read(path)
+        .unwrap();
+    let jar_bytes = bytes::Bytes::from(jar_buf);
+    let jar_hash = sha1_async(jar_bytes).await;
+
+    let json_metadata = &extract_json_from_jar(path).expect("Failed to extract JSON from JAR");
+
+    let mod_name = json_metadata["name"].as_str().unwrap().to_string(); 
+    let mod_description = json_metadata["description"].as_str().unwrap().to_string(); 
+    let timestamp_added = &get_file_datetime(path).expect("Failed to extract timestamp.");
+    
+    Ok(JARLoadedMod {
+        file_hash: jar_hash,
+        absolute_path: file_path,
+        timestamp_added: *timestamp_added,
+        mod_name: mod_name,
+        description: mod_description,
+    })
+}
+
+// Get timestamp of the file being modified.
+fn get_file_datetime(file_path: &Path) -> Result<SystemTime, String> {
+    let metadata = fs::metadata(file_path).map_err(|err| err.to_string())?;
+    let modified_time = metadata.modified().map_err(|err| err.to_string())?;
+    Ok(modified_time)
+}
+
+// Extract JSON from JAR.
+// todo: this always looks for the first JSON file. Is it guaranteed that the meta-data file will always have the same name as the mod? 
+fn extract_json_from_jar(jar_file: &Path) -> Result<Value, String> {
+    let file = File::open(jar_file).map_err(|e| e.to_string())?;
+    let mut archive = ZipArchive::new(BufReader::new(file)).map_err(|e| e.to_string())?;
+
+    for i in 0..archive.len() {
+        let file = archive.by_index(i).map_err(|e| e.to_string())?;
+        let file_name = file.name();
+        if file_name.ends_with(".json") {
+            let content: Value = serde_json::from_reader(file).expect("JSON was not well-formatted");
+            return Ok(content);
+        }
+    }
+    return Err("No JSON files found!".to_string());
+}
```

Comment:
This won't work on different mod loaders- only Fabric. Forge uses TOML for metadata which is stored at `META-INF/mods.toml`. Quilt has a different format.

Reply:
Fixed